### PR TITLE
feat: onboard intercom v2 destination

### PR DESCRIPTION
**Pull request by manish339k**
## What are the changes introduced in this PR?

Onboarding intercom v2 destination. It supports Oauth and api version 2.10

## What is the related Linear task?

Resolves INT-2474

## Please explain the objectives of your changes below

Put down any required details on the broader aspect of your changes. If there are any dependent changes, **mandatorily** mention them here

### Any changes to existing capabilities/behaviour, mention the reason & what are the changes ?

N/A

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

N/A

### Any technical or performance related pointers to consider with the change?

N/A

@coderabbitai review

<hr>

### Developer checklist

- [ ] My code follows the style guidelines of this project

- [ ] **No breaking changes are being introduced.**

- [ ] All related docs linked with the PR?

- [ ] All changes manually tested?

- [ ] Any documentation changes needed with this change?

- [ ] Is the PR limited to 10 file changes?

- [ ] Is the PR limited to one linear task?

- [ ] Are relevant unit and component test-cases added in **new readability format**?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.

### CODE REVIEW

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/3721/test-report.html

File: src/v0/destinations/intercom_v2/transform.js [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+const process = async (event) => {
+  const { destination } = event;
+  const apiVersion = getApiVersion(destination);
+  let response;
+  if (apiVersion === 'v2') {
+    response = await processEvent(event);
+  } else {
+    throw new InstrumentationError(`apiVersion ${apiVersion} is not supported.`);
+  }
+  return response;
```
Comment (utsabc) [inline src/v0/destinations/intercom_v2/transform.js:162]: If this is a new destination we should not give option of Anything apart from `v2` hence at this point this field in UI config doesnt make sense

File: src/v0/destinations/intercom_v2/utils.js [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+    endpoint,
+    finalPayload,
+    {
+      headers,
+    },
+    statTags,
+  );
+
+  if (!isHttpStatusSuccess(response.status)) {
+    intercomErrorHandler('Unable to Create or Update Company due to', response);
```
Comment (utsabc) [inline src/v0/destinations/intercom_v2/utils.js:198]: What about Oauth scenario handling ?

File: src/v0/destinations/intercom_v2/utils.js [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+        {
+          headers,
+        },
+        statTags,
+      );
+
+      if (!isHttpStatusSuccess(response.status)) {
+        intercomErrorHandler('Unable to Add or Update the Tag to Company due to', response);
+      }
+    }),
```
Comment (utsabc) [inline src/v0/destinations/intercom_v2/utils.js:274]: Need to handle the Oauth response for these calls

File: src/v0/destinations/intercom_v2/utils.js [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+  };
+
+  const { processedResponse: response } = await handleHttpRequest(
+    'POST',
+    endpoint,
+    data,
+    {
+      headers,
+    },
+    statTags,
```
Comment (ItsSudip) [inline src/v0/destinations/intercom_v2/utils.js:138]: Can we send the metadata to `handleHttpRequest` function calls? It will help us with logging.

Comment (manish339k) [reply src/v0/destinations/intercom_v2/utils.js:138]: metadata is present in statTags.

File: src/v0/destinations/intercom_v2/data/IntercomTrackConfig.json [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    "destKey": "created_at",
+    "sourceKeys": "timestamp",
+    "sourceFromGenericMap": true,
+    "metadata": {
+      "type": "secondTimestamp"
+    }
+  },
+  {
+    "destKey": "id",
+    "sourceKeys": ["message.properties.id", "message.traits.id"]
```
[the remark this answers is not in the document]
Comment (manish339k) [reply src/v0/destinations/intercom_v2/data/IntercomTrackConfig.json:27]: this id is different, it is id provided by intercom. external_id is equivalent to user_id in track call and we are taking user_id from GenericMap.

File: src/v0/destinations/intercom_v2/networkHandler.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+} = require('../../../adapters/utils/networkUtils');
+const { AUTH_STATUS_INACTIVE } = require('../../../adapters/networkhandler/authConstants');
+const { prepareProxyRequest, proxyRequest } = require('../../../adapters/network');
+const { TransformerProxyError } = require('../../util/errorTypes');
+const tags = require('../../util/tags');
+
+// ref: https://github.com/intercom/oauth2-intercom
+// Intercom's OAuth implementation does not use refresh tokens. Access tokens are valid until a user revokes access manually, or until an app deauthorizes itself.
+const getAuthErrCategory = (status) => {
+  if (status === 401) {
```
[the remark this answers is not in the document]
Comment (manish339k) [reply src/v0/destinations/intercom_v2/networkHandler.js:15]: for 401 status, if access token is revoked or wrong, in both cases we need a new token. https://developers.intercom.com/docs/references/2.10/rest-api/errors/http-responses

File: src/v0/destinations/intercom_v2/utils.js [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+ * method to handle error during api call
+ * ref docs: https://developers.intercom.com/docs/references/rest-api/errors/http-responses/
+ * e.g.
+ * 400 - code: parameter_not_found (or parameter_invalid), message: company not specified
+ * 401 - code: unauthorized, message: Access Token Invalid
+ * 404 - code: company_not_found, message: Company Not Found
+ * @param {*} message
+ * @param {*} processedResponse
+ */
+const intercomErrorHandler = (message, processedResponse) => {
```
[the remark this answers is not in the document]
Comment (manish339k) [reply src/v0/destinations/intercom_v2/utils.js:41]: Error response format is different for delivery and intermediate api call.

File: src/v0/destinations/intercom_v2/utils.js (hunk 4 above)
Comment (manish339k) [reply src/v0/destinations/intercom_v2/utils.js:138]: metadata coming in log. resolving this.

File: src/v0/destinations/intercom_v2/data/IntercomTrackConfig.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (manish339k) [reply src/v0/destinations/intercom_v2/data/IntercomTrackConfig.json:27]: done.

File: src/v0/destinations/intercom_v2/transform.js [hunk 8]
```diff
@@ -48,6 +49,10 @@ const transformTrackPayload = (event) => {
   if (!(payload.user_id || payload.email)) {
     throw new InstrumentationError('Either email or userId is required for Track call');
   }
+  if (!payload.id) {
+    const integrationsObj = getIntegrationsObj(message, 'INTERCOM');
+    payload.id = integrationsObj?.id;
+  }
```
Comment (ItsSudip) [inline src/v0/destinations/intercom_v2/transform.js:55]: If the customer has provided intercom id through integrations object, we can use that to make the track call. We can skip the check for email, userId.

**Review by utsabc — APPROVED** [review]

**Review by ItsSudip — APPROVED** [review]

**Review by utsabc — APPROVED** [review]

**Pull request merged**